<a href="https://colab.research.google.com/github/Auvarmf/BasicMachineLearning/blob/main/PERTEMUAN5_AuvarMahsaFahlevi_2117051027.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Pertemuan 5

## Tahapan Machine Learning dalam hal Coding


---

Tahapannya:

*   Preprocessing
*   Modeling
*   Evaluasi

Ketika data sudah sangan kompleks maka diperlukan tahapan preprocessing, salah satu contoh tahapan preprocessing adalah menghapus NaN Values.

Data-data yang tidak bagus harus melewati tahapan preprocessing agar output data yang dihasilkan lebih baik atau lebih masuk akal.

Melakukan pemodelan atau training data.

Melakukan evaluasi dengan cara mencari nilai akurasi dari sebuah data.



## Library

In [134]:
import pandas as pd
import numpy as np
import sklearn 
from sklearn.cluster import AgglomerativeClustering
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score

## Latihan 1

## Dataset

In [135]:
# buat data tabular seperti dibawah ini
df = pd.DataFrame(
    {
        "ID Buah" : pd.Series([1, 2, 3, 4, 5]),
        "Diameter" : pd.Series([7.0, 7.1, 6.5, 6.6, 6.8]),
        "Berat" : pd.Series([165, 170, 180, 195, 200]),
    }
)
df

ID Buah  Diameter  Berat
0        1       7.0    165
1        2       7.1    170
2        3       6.5    180
3        4       6.6    195
4        5       6.8    200

## Preprocessing

In [136]:
# pemilihan feature
X = df.drop("ID Buah", axis = 1)
X

Diameter  Berat
0       7.0    165
1       7.1    170
2       6.5    180
3       6.6    195
4       6.8    200

## Modeling
Menggunakan algoritma agglomerative clustering

In [137]:
clustering = AgglomerativeClustering().fit(X)
# hasil clustering
df["kluster"] = clustering.labels_
df

ID Buah  Diameter  Berat  kluster
0        1       7.0    165        0
1        2       7.1    170        0
2        3       6.5    180        0
3        4       6.6    195        1
4        5       6.8    200        1

## Latihan 2

## Dataset

In [138]:
df2 = pd.read_csv("https://raw.githubusercontent.com/rahmantaufik27/dataset/main/study/student_score_pass.csv")
df2

npm  activeness_score  assignment_score  pre_test_score  \
0    2022041101                99              71.0            93.0   
1    2022041102                96              77.0             NaN   
2    2022041103                86              71.0            84.0   
3    2022041104                81              96.0             NaN   
4    2022041105                93              85.0            68.0   
..          ...               ...               ...             ...   
367  2022041468                91               NaN            31.0   
368  2022041469                96              87.0            55.0   
369  2022041470               569              83.0            89.0   
370  2022041471                84              92.0            87.0   
371  2022041472                90              99.0            96.0   

     post_test_score  midterm_test_score  end_semester_test_score pass_test  
0               52.0                44.0                       24  not pass  
1               27.0               100.0                       89      pass  
2               17.0                21.0                       41  not pass  
3                NaN                11.0                       99  not pass  
4               29.0                44.0                       50  not pass  
..               ...                 ...                      ...       ...  
367             52.0                31.0                       46  not pass  
368             56.0                95.0                       11      pass  
369             15.0                86.0                       13  not pass  
370             97.0                42.0                       27      pass  
371             60.0                82.0                       65  not pass  

[372 rows x 8 columns]

## Preprocessing

In [139]:
# pemilihan feature
df_training = df2.drop(["npm", "pass_test"], axis = 1)
# hapus NaN values
df_training = df_training.dropna()
df_training

activeness_score  assignment_score  pre_test_score  post_test_score  \
0                  99              71.0            93.0             52.0   
2                  86              71.0            84.0             17.0   
4                  93              85.0            68.0             29.0   
5                  85              95.0            34.0             73.0   
6                  95              79.0            52.0             54.0   
..                ...               ...             ...              ...   
366                88              75.0            80.0             11.0   
368                96              87.0            55.0             56.0   
369               569              83.0            89.0             15.0   
370                84              92.0            87.0             97.0   
371                90              99.0            96.0             60.0   

     midterm_test_score  end_semester_test_score  
0                  44.0                       24  
2                  21.0                       41  
4                  44.0                       50  
5                  57.0                       98  
6                  85.0                       13  
..                  ...                      ...  
366                55.0                       82  
368                95.0                       11  
369                86.0                       13  
370                42.0                       27  
371                82.0                       65  

[312 rows x 6 columns]

## Modeling

In [140]:
model = AgglomerativeClustering().fit(df_training)
# hasil clustering
df_training["kluster"] = model.labels_
df_training

activeness_score  assignment_score  pre_test_score  post_test_score  \
0                  99              71.0            93.0             52.0   
2                  86              71.0            84.0             17.0   
4                  93              85.0            68.0             29.0   
5                  85              95.0            34.0             73.0   
6                  95              79.0            52.0             54.0   
..                ...               ...             ...              ...   
366                88              75.0            80.0             11.0   
368                96              87.0            55.0             56.0   
369               569              83.0            89.0             15.0   
370                84              92.0            87.0             97.0   
371                90              99.0            96.0             60.0   

     midterm_test_score  end_semester_test_score  kluster  
0                  44.0                       24        0  
2                  21.0                       41        0  
4                  44.0                       50        0  
5                  57.0                       98        0  
6                  85.0                       13        0  
..                  ...                      ...      ...  
366                55.0                       82        0  
368                95.0                       11        0  
369                86.0                       13        1  
370                42.0                       27        0  
371                82.0                       65        0  

[312 rows x 7 columns]

# Tugas 

Evaluasi model algortima agglomerative clustering dari data latihan 2 dengan mencari nilai akurasinya

Hint: splitting data, modeling, accuracy score

Nilai Akurasi: 0.5079365079365079

In [141]:
# melakukan pengambilan data
df = pd.read_csv("https://raw.githubusercontent.com/rahmantaufik27/dataset/main/study/student_score_pass.csv")
df

npm  activeness_score  assignment_score  pre_test_score  \
0    2022041101                99              71.0            93.0   
1    2022041102                96              77.0             NaN   
2    2022041103                86              71.0            84.0   
3    2022041104                81              96.0             NaN   
4    2022041105                93              85.0            68.0   
..          ...               ...               ...             ...   
367  2022041468                91               NaN            31.0   
368  2022041469                96              87.0            55.0   
369  2022041470               569              83.0            89.0   
370  2022041471                84              92.0            87.0   
371  2022041472                90              99.0            96.0   

     post_test_score  midterm_test_score  end_semester_test_score pass_test  
0               52.0                44.0                       24  not pass  
1               27.0               100.0                       89      pass  
2               17.0                21.0                       41  not pass  
3                NaN                11.0                       99  not pass  
4               29.0                44.0                       50  not pass  
..               ...                 ...                      ...       ...  
367             52.0                31.0                       46  not pass  
368             56.0                95.0                       11      pass  
369             15.0                86.0                       13  not pass  
370             97.0                42.0                       27      pass  
371             60.0                82.0                       65  not pass  

[372 rows x 8 columns]

In [142]:
# memilihan feature
# pada kasus ini kita akan menghapus data npm
df_training = df.drop(["npm"], axis = 1)

# konversi nilai pada kolom "pass_test" menjadi integer
# karena kita akan menggunakan data pass_test, maka dilakukan konversi terlebih dahulu
# hal ini dilakukan karena data pass_test merupakan string sehingga akan terjai error saat di olah, sehingga harus dilakukan konversi terlebih dahulu
df_training["pass_test"] = df_training["pass_test"].apply(lambda x: 1 if x == "pass" else 0)

# menghapus NaN values
df_training = df_training.dropna()
df_training

activeness_score  assignment_score  pre_test_score  post_test_score  \
0                  99              71.0            93.0             52.0   
2                  86              71.0            84.0             17.0   
4                  93              85.0            68.0             29.0   
5                  85              95.0            34.0             73.0   
6                  95              79.0            52.0             54.0   
..                ...               ...             ...              ...   
366                88              75.0            80.0             11.0   
368                96              87.0            55.0             56.0   
369               569              83.0            89.0             15.0   
370                84              92.0            87.0             97.0   
371                90              99.0            96.0             60.0   

     midterm_test_score  end_semester_test_score  pass_test  
0                  44.0                       24          0  
2                  21.0                       41          0  
4                  44.0                       50          0  
5                  57.0                       98          0  
6                  85.0                       13          0  
..                  ...                      ...        ...  
366                55.0                       82          0  
368                95.0                       11          1  
369                86.0                       13          0  
370                42.0                       27          1  
371                82.0                       65          0  

[312 rows x 7 columns]

In [143]:
# memisahkan data menjadi data latih dan data uji
X_train, X_test, y_train, y_test = train_test_split(df_training.drop(["pass_test"], axis=1), df_training["pass_test"], test_size=0.2, random_state=2)

# random_state adalah argumen opsional yang digunakan untuk menentukan seed atau biji secara acak yang digunakan dalam pembagian data.
# test_size digunakan untuk menentukan ukuran data test sebagai proporsi dari keseluruhan dataset.
# test_size = 0.2, maka 20% dari data akan digunakan sebagai data test, sementara 80% sisanya akan digunakan sebagai data training.
# ukuran test_size bisa disesuaikan dengan data yang dimiliki.

In [144]:
# melakukan modeling Agglomerative Clustering
model = AgglomerativeClustering().fit(X_train)
model

AgglomerativeClustering()

In [145]:
# melakukan prediksi pada data latih
train_prediksi = model.fit_predict(X_train)
train_prediksi

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0])

In [146]:
# melakukan prediksi pada data uji
test_prediksi = model.fit_predict(X_test)
test_prediksi

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])

In [147]:
# menghitung nilai akurasi data latih
train_akurasi = accuracy_score(y_train, train_prediksi)
train_akurasi

0.642570281124498

In [148]:
# menghitung nilai akurasi data uji
test_akurasi = accuracy_score(y_test, test_prediksi)
test_akurasi

0.5873015873015873

Akurasi data tersebut sesuai dengan yang diharapkan, akurasi data tersebut bisa ditingkatkan lagi menjadi lebih tinggi dengan Random Forest Classifier

In [149]:
# model yang digunakan adalah Random Forest 
# Random Forest Classifier digunakan dengan 500 estimators (jumlah pohon keputusan yang dibangun) dan random state sebesar 2.
model = RandomForestClassifier(n_estimators=510, random_state=2)
model

RandomForestClassifier(n_estimators=510, random_state=2)

In [150]:
# melatih model menggunakan data train
model.fit(X_train, y_train)

# melakukan prediksi pada data test
y_prediksi = model.predict(X_test)
y_prediksi

array([1, 1, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0,
       0, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1,
       1, 0, 0, 0, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0])

In [151]:
# menghitung akurasi pada data test
akurasi = accuracy_score(y_test, y_prediksi)

# mengubah akurasi ke percent
akurasi_percent = akurasi * 100
print("Akurasi :", akurasi_percent, "%")

Akurasi : 80.95238095238095 %


Nilai akurasi yang bisa saya dapatkan setelah menggunakan Random Forest Classifier adalah 80.95238095238095 %